sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3088,7 +3088,12 @@ class Analyzer(override val catalogManager: CatalogManager)
         val projection = TableOutputResolver.resolveOutputColumns(
           v2Write.table.name, v2Write.table.output, v2Write.query, v2Write.isByName, conf)
         if (projection != v2Write.query) {
-          v2Write.withNewQuery(projection)
+          val cleanedTable = v2Write.table match {
+            case r: DataSourceV2Relation =>
+              r.copy(output = r.output.map(CharVarcharUtils.cleanAttrMetadata))
```

Comment:
We remove the char/varchar metadata after length check expressions are added, so that we don't do it repeatedly and this rule is idempotent.

Reply:
No it doesn't. Metadata is fine as it's harmless. We only need to watch out for some specific rules that look at the char/varchar metadata, and make sure they are idempotent.

As a fact, the added cast and length check expression is wrapped by an `Alias` which retains char/varchar metadata. So the output attributes of `Project` above the v2 relation still have metadata. It's necessary as we need to rely on it later to do padding for char type column comparison.